nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/pgp/PGPControllerService.java

```diff
@@ -0,0 +1,893 @@
+import org.bouncycastle.openpgp.PGPLiteralData;
+import org.bouncycastle.openpgp.PGPLiteralDataGenerator;
+import org.bouncycastle.openpgp.PGPObjectFactory;
+import org.bouncycastle.openpgp.PGPPBEEncryptedData;
+import org.bouncycastle.openpgp.PGPPrivateKey;
+import org.bouncycastle.openpgp.PGPPublicKey;
+import org.bouncycastle.openpgp.PGPPublicKeyEncryptedData;
+import org.bouncycastle.openpgp.PGPPublicKeyRing;
+import org.bouncycastle.openpgp.PGPSecretKey;
+import org.bouncycastle.openpgp.PGPSecretKeyRing;
+import org.bouncycastle.openpgp.PGPSecretKeyRingCollection;
+import org.bouncycastle.openpgp.PGPSignature;
+import org.bouncycastle.openpgp.PGPSignatureGenerator;
+import org.bouncycastle.openpgp.PGPSignatureList;
+import org.bouncycastle.openpgp.jcajce.JcaPGPObjectFactory;
+import org.bouncycastle.openpgp.jcajce.JcaPGPPublicKeyRingCollection;
+import org.bouncycastle.openpgp.operator.KeyFingerPrintCalculator;
+import org.bouncycastle.openpgp.operator.PBEDataDecryptorFactory;
+import org.bouncycastle.openpgp.operator.PBESecretKeyDecryptor;
+import org.bouncycastle.openpgp.operator.PGPDigestCalculatorProvider;
+import org.bouncycastle.openpgp.operator.PublicKeyDataDecryptorFactory;
+import org.bouncycastle.openpgp.operator.bc.BcKeyFingerprintCalculator;
+import org.bouncycastle.openpgp.operator.bc.BcPGPDigestCalculatorProvider;
+import org.bouncycastle.openpgp.operator.jcajce.JcaPGPContentSignerBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcaPGPContentVerifierBuilderProvider;
+import org.bouncycastle.openpgp.operator.jcajce.JcaPGPDigestCalculatorProviderBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePBEDataDecryptorFactoryBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePBEKeyEncryptionMethodGenerator;
+import org.bouncycastle.openpgp.operator.jcajce.JcePBESecretKeyDecryptorBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePGPDataEncryptorBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyDataDecryptorFactoryBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyKeyEncryptionMethodGenerator;
+import org.bouncycastle.util.io.Streams;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.security.Security;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.Deflater;
+
+
+/**
+ * This class defines an implementation of {@link PGPService} suitable for use with the PGP processors in this package.
+ *
+ * In addition to basic Controller Service functionality this class exposes high-level, stream-based cryptographic operations
```

Comment:
I think I was unclear in our previous conversations. I was imagining an (internal) service which provided these high-level functions, separate from the NiFi `interface SpecificControllerService` and `class StandardSpecificControllerService implements SpecificControllerService` (see `SSLContextService` and `StandardSSLContextService`/`RestrictedStandardSSLContextService` or `KerberosCredentialsService`/`KeytabCredentialsService` for examples). 

This would provide the following benefits:
* The `PGPService` would provide easy-to-consume and safe-to-use PGP operations to NiFi code. 
* The PGP-specific logic would only be here, making upgrades/change-of-library/etc. work isolated and contained. 
* The `PGPKeyMaterialService` (example name) would have a `StandardPGPKeyMaterialService` implementation which defined the necessary `PropertyDescriptors` and logic for parsing the low-level values out of the NiFi concepts of PD/`AllowableValue` into Java types. These parameters to the cryptographic operations I believe you have encapsulated in the various `EncryptOptions`, etc. classes. I would expect these to share a common marker interface like `PGPOptions`. 
* The parsed parameters would be extracted from the controller service implementation within the respective processor implementations in `#onTrigger()` and passed to the `PGPService` instance to perform the actual cryptographic operations. In my view, controller services and processors should both be minimal in the sense that they are wrappers/adaptors for fitting the actual behavior/logic into the NiFi framework, while the logic, especially unrelated to NiFi-specific concepts, should be encapsulated on its own to provide clear separation of concerns. This is demonstrated (incompletely and not as cleanly as I would like) in `EncryptContent` which delegates the actual operations to `OpenPGPKeyBasedEncryptor`, `OpenPGPPasswordBasedEncryptor`, `KeyBasedEncryptor`, and `PasswordBasedEncryptor`.

Reply:
This will require a few hours to re-organize, but I don't foresee any real difficulty.  The controller class is clearly doing double-duty and the behaviors are cleanly marked and clearly separable.  If history is any guide, the packaging and module-making parts will be the most difficult.  Do you have any suggested module layout in mind?

Reply:
For now I've refactored these methods into classes/interfaces into a new package `org.apache.nifi.security.pgp`; please advise on that choice when you can.

Reply:
I think that package is appropriate. The pure cryptographic services can go in the `nifi-security-utils` module. The module already includes BouncyCastle as a dependency so you should not need to modify `LICENSE` or `NOTICE`.

Reply:
I've extracted the (internal) service you mentioned to an interface called `PGPOperator` and a standard implementation called `StandardPGPOperator`, both in package `org.apache.nifi.security.pgp` and module `nifi-security-utils`.

I picked "operator" to avoid confusion with "service" and "controller".  "PGPUtil" or similar wasn't considered because we have (at least) two of those already.

NB the package placement of the Controller Service interfaces; this was intentional but I'm not certain if it's optimal.